### PR TITLE
Hack/explicit bundles

### PR DESCRIPTION
**Pull request by Mic92**
The pull request based on https://github.com/docker/docker/pull/12376. I have created a  seperate pull request, because this one actually adds new functionality. See "hack: explicit bundles for integration-cli prequisites" for details.

### CODE REVIEW

**Comment by jfrazelle** [comment]
needs rebase sorry :(

**Comment by Mic92** [comment]
@jfrazelle done

**Comment by jfrazelle** [comment]
oh man so sorry needs another rebase

**Comment by Mic92** [comment]
@jfrazelle done

**Comment by tianon** [comment]
Generally a fan, although I'm not sure I like that it changes `DEST` for each sub-bundle, or that the bundles are renamed.  Bundles that start with `.` shouldn't be a problem, especially since they're still not bundles that ought to be run in isolation unless you know what you're doing (which is why they started with `.` to begin with).

**Comment by Mic92** [comment]
The sub-bundles seems to create sub directories in `$DEST`, so this duplication can be removed just using $DEST instead. A different solution would be a conditional check.

**Comment by tianon** [comment]
Here's what I meant more concretely (which I'm happy to just submit as a PR to your branch if you'd like):
```patch
diff --git a/hack/make.sh b/hack/make.sh
index b9c2aef..eb25e22 100755
--- a/hack/make.sh
+++ b/hack/make.sh
@@ -264,11 +264,9 @@ hash_files() {
 }
 
 bundle() {
-	bundlescript=$1
-	bundle=$(basename $bundlescript)
-	echo "---> Making bundle: $bundle (in bundles/$VERSION/$bundle)"
-	mkdir -p "bundles/$VERSION/$bundle"
-	source "$bundlescript" "$(pwd)/bundles/$VERSION/$bundle"
+	local bundle="$1"
+	echo "---> Making bundle: $(basename "$bundle") (in $DEST)"
+	source "$SCRIPTDIR/make/$bundle"
 }
 
 main() {
@@ -292,7 +290,9 @@ main() {
 		bundles=($@)
 	fi
 	for bundle in ${bundles[@]}; do
-		bundle "$SCRIPTDIR/make/$bundle"
+		export DEST="bundles/$VERSION/$(basename "$bundle")"
+		mkdir -p "$DEST"
+		bundle "$bundle"
 		echo
 	done
 }
```

(and then every use of `bundle "$SCRIPTDIR/make/..."` replaced with just `bundle '...'`)

```patch
diff --git a/hack/make/binary b/hack/make/binary
index d3ec293..9642f51 100644
--- a/hack/make/binary
+++ b/hack/make/binary
@@ -1,7 +1,6 @@
 #!/bin/bash
 set -e
 
-DEST=$1
 BINARY_NAME="docker-$VERSION"
 BINARY_EXTENSION="$(binary_extension)"
 BINARY_FULLNAME="$BINARY_NAME$BINARY_EXTENSION"
diff --git a/hack/make/build-deb b/hack/make/build-deb
index 36e1177..7acaa4b 100644
--- a/hack/make/build-deb
+++ b/hack/make/build-deb
@@ -1,8 +1,6 @@
 #!/bin/bash
 set -e
 
-DEST=$1
-
 # subshell so that we can export PATH without breaking other things
 (
 	source "${MAKEDIR}/.integration-daemon-start"
diff --git a/hack/make/build-rpm b/hack/make/build-rpm
index 0f3ff6d..3a7bfd4 100644
--- a/hack/make/build-rpm
+++ b/hack/make/build-rpm
@@ -1,8 +1,6 @@
 #!/bin/bash
 set -e
 
-DEST=$1
-
 # subshell so that we can export PATH without breaking other things
 (
 	source "$(dirname "$BASH_SOURCE")/.integration-daemon-start"
diff --git a/hack/make/cover b/hack/make/cover
index ca772d0..624943b 100644
--- a/hack/make/cover
+++ b/hack/make/cover
@@ -1,8 +1,6 @@
 #!/bin/bash
 set -e
 
-DEST="$1"
-
 bundle_cover() {
 	coverprofiles=( "$DEST/../"*"/coverprofiles/"* )
 	for p in "${coverprofiles[@]}"; do
diff --git a/hack/make/cross b/hack/make/cross
index 368ebc5..fcd4ef6 100644
--- a/hack/make/cross
+++ b/hack/make/cross
@@ -1,8 +1,6 @@
 #!/bin/bash
 set -e
 
-DEST=$1
-
 # explicit list of os/arch combos that support being a daemon
 declare -A daemonSupporting
 daemonSupporting=(
diff --git a/hack/make/dynbinary b/hack/make/dynbinary
index e1b65b4..e5fc0be 100644
--- a/hack/make/dynbinary
+++ b/hack/make/dynbinary
@@ -1,8 +1,6 @@
 #!/bin/bash
 set -e
 
-DEST=$1
-
 if [ -z "$DOCKER_CLIENTONLY" ]; then
 	source "${MAKEDIR}/.dockerinit"
 
diff --git a/hack/make/dyngccgo b/hack/make/dyngccgo
index 7bdd404..df76ac7 100644
--- a/hack/make/dyngccgo
+++ b/hack/make/dyngccgo
@@ -1,8 +1,6 @@
 #!/bin/bash
 set -e
 
-DEST=$1
-
 if [ -z "$DOCKER_CLIENTONLY" ]; then
 	source "${MAKEDIR}/.dockerinit-gccgo"
 
diff --git a/hack/make/gccgo b/hack/make/gccgo
index 896c2d4..972934a 100644
--- a/hack/make/gccgo
+++ b/hack/make/gccgo
@@ -1,7 +1,6 @@
 #!/bin/bash
 set -e
 
-DEST=$1
 BINARY_NAME="docker-$VERSION"
 BINARY_EXTENSION="$(binary_extension)"
 BINARY_FULLNAME="$BINARY_NAME$BINARY_EXTENSION"
diff --git a/hack/make/test-docker-py b/hack/make/test-docker-py
index ac5ef35..cbd59ad 100644
--- a/hack/make/test-docker-py
+++ b/hack/make/test-docker-py
@@ -1,8 +1,6 @@
 #!/bin/bash
 set -e
 
-DEST=$1
-
 # subshell so that we can export PATH without breaking other things
 (
 	source "${MAKEDIR}/.integration-daemon-start"
diff --git a/hack/make/test-integration-cli b/hack/make/test-integration-cli
index db1cb29..61d62f1 100644
--- a/hack/make/test-integration-cli
+++ b/hack/make/test-integration-cli
@@ -1,8 +1,6 @@
 #!/bin/bash
 set -e
 
-DEST=$1
-
 bundle_test_integration_cli() {
 	go_test_dir ./integration-cli
 }
diff --git a/hack/make/test-unit b/hack/make/test-unit
index 7b6ce08..6b85863 100644
--- a/hack/make/test-unit
+++ b/hack/make/test-unit
@@ -1,7 +1,6 @@
 #!/bin/bash
 set -e
 
-DEST=$1
 : ${PARALLEL_JOBS:=$(nproc 2>/dev/null || echo 1)} # if nproc fails (usually because we don't have it), let's not parallelize by default
 
 RED=$'\033[31m'
diff --git a/hack/make/tgz b/hack/make/tgz
index fa297e1..a512b83 100644
--- a/hack/make/tgz
+++ b/hack/make/tgz
@@ -1,6 +1,5 @@
 #!/bin/bash
 
-DEST="$1"
 CROSS="$DEST/../cross"
 
 set -e
diff --git a/hack/make/ubuntu b/hack/make/ubuntu
index 7543789..d3c952d 100644
--- a/hack/make/ubuntu
+++ b/hack/make/ubuntu
@@ -1,7 +1,5 @@
 #!/bin/bash
 
-DEST=$1
-
 PKGVERSION="${VERSION//-/'~'}"
 # if we have a "-dev" suffix or have change in Git, let's make this package version more complex so it works better
 if [[ "$VERSION" == *-dev ]] || [ -n "$(git status --porcelain)" ]; then
```

**Comment by icecrime** [comment]
@Mic92 Sorry, needs another rebase :-(

**Comment by Mic92** [comment]
@icecrime done

**Comment by Mic92** [comment]
Now it is working.

**Comment by icecrime** [comment]
Ping @jfrazelle @tianon!

**Comment by tianon** [comment]
> Generally a fan, although I'm not sure I like that [...] the bundles are renamed.  Bundles that start with `.` shouldn't be a problem, especially since they're still not bundles that ought to be run in isolation unless you know what you're doing (which is why they started with `.` to begin with).

Also, the double quotes that don't use interpolation should be single quotes instead:
```bash
bundle '.test-integration-start'
# instead of
bundle ".test-integration-start"
```

File: hack/make/.ensure-emptyfs [hunk 1]
```diff
@@ -1,4 +1,5 @@
 #!/bin/bash
+(
```
Comment (tianon) [inline hack/make/.ensure-emptyfs]: What's the reasoning behind adding an explicit subshell to each of these?

File: hack/make/.integration-daemon-setup [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+bundle .ensure-emptyfs
+bundle .ensure-frozen-images
+bundle .ensure-httpserver
```
Comment (tianon) [inline hack/make/.integration-daemon-setup]: This is only used in one place, right?  With the simpler `bundle ...` invocation, it seems like it's not worth splitting out IMO.

File: hack/make/build-deb [hunk 3]
```diff
@@ -63,5 +62,5 @@ set -e
 		docker rmi "$tempImage"
 	done
 
-	source "${MAKEDIR}/.integration-daemon-stop"
+	bundle bundle .test-integration-stop
```
Comment (tianon) [inline hack/make/build-deb]: `s/bundle bundle/bundle/`

**Comment by tianon** [comment]
Looking good, just a few minor nits! :+1:

File: hack/make/.integration-daemon-setup (hunk 2 above)
Comment (Mic92) [inline hack/make/.integration-daemon-setup]: The original idea was to have an one file to source to start an interactive shell, where one can start and debug single integration tests.

File: hack/make/.ensure-emptyfs (hunk 1 above)
Comment (Mic92) [inline hack/make/.ensure-emptyfs]: In the first place, they had exported environment variables.

**Comment by thaJeztah** [comment]
ping @tianon @jfrazelle PTAL

File: hack/make/.integration-daemon-setup (hunk 2 above)
Comment (tianon) [inline hack/make/.integration-daemon-setup]: Ah, ok; that's fair! :+1:

**Comment by tianon** [comment]
LGTM cc @jfrazelle

**Comment by jfrazelle** [comment]
LGTM

**Pull request merged**